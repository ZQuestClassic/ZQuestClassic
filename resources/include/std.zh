// std.zh
// 28th December, 2020 for ZC 2.55 Alpha 86
// Standard header for ZScript.
// Please don't edit this file. Instead, create a new header file for your
// quests' scripts,  add constants and procedures to that file,
// and #include it separately.

//Options
#option BINARY_32BIT off
	//Would break a number of things if set on
#option SHORT_CIRCUIT on
	//Should work for everything in std_zh, and just speed it up

#includepath "include"

//Version Metadata
#include "std_zh/std_meta.zh"

//User Configuration Files
#include "std_zh/std_user_defs.zh" //Typedefs: Always first.
#include "std_zh/std.cfg"

//Main Files
#include "std_zh/limits.zh"
#include "std_zh/std_constants.zh"
#include "std_zh/ffrules.zh"
#include "std_zh/std_functions.zh"
#include "std_zh/math.zh"
#include "std_zh/weapon.zh"
#include "std_zh/std_sideview.zh" //Sideview and platforming

//ZScript string handling functions, akin to those found in C
//#include "string.zh"
#include "std_zh/string/string_constants.zh" //Now #includeed from here. This prevents conflicts for those people
#include "std_zh/string/string_functions.zh" //who would #include string.zh from a secondary source, as that file is now 
			     //empty, and the string.zh components load from here. 

//Keyboard handling for 2.54 and above
#include "std_zh/std_keyboard.zh"

//Locale
#include "std_zh/locale/std_EnglishIntl.zh" //International English Localisation

//Legacy Wrappers
#include "std_zh/std_legacy.zh"