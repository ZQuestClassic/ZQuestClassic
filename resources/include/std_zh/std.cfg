//Settings

const int STD_FORCE_OLD_DISTANCE = 1;	//If enabled, Distance() will use Sqrt() instead of SafeSqrt().
					//! If you use this, disable the quest rule 'Log Script Errors to Allegro.log
					//! before distributing your quest!
					
const int STD_DISTANCE_USE_SAFESQRT = 0; 	//If enebaled, Distance will use SafeSqrt(). Otherwise, it will return a factor, 
					
const int STD_EMULATE_ENGINE_COLLISION = 0; // If enabled, Collision hitboxs will more precisely emulate the
					    // internal collision. This differs to how scripted collision
					    // has always worked.
					    

const int STD_CENTERXY_USES_HITBOXES = 0; // If enabled, CenterX() and CenterY() will use object hitboxes
					  // and offsets by default, rather than sprite sizes when calculating
					  // the centre of objects.
					  

const int STD_NEVER_USE_SAFESQRT = 1; //Enable to force any calls to SafeSqrt() to use Sqrt() instead.

const int MAX_HIGHEST_LEVEL_ITEM_CHECK = 255; //The highest item ID checked by GetHighestLevelItem(). 

const int REVISED_ROTATEPOINT_X_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointX()
const int REVISED_ROTATEPOINT_Y_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointY()


const int MAX_NOACTION_INPUT = 14; //Checks A, B, L, R, U/D/L/R, Ex1, Ex2, Ex3, Ex4
	//Use CB_MAX to force the analogue stick false. 
const int NOACTION_SKIP_MAP = 0; //If enabled, NoAction will not disable pressing the Map button
const int NOACTION_SKIP_START = 0; //If enabled, NoAction will not disable pressing the Start button

const int OLD_ONSIDEPLATFORM = 0; //Change to 1 to use old OnSidePlatform calls.

const int OLD_RAND_FUNCS = 0; //Change to 1 to use old Rand() logic
