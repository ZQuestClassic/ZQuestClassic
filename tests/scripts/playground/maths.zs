#include "std.zh"

ffc script Maths
{
    void run()
    {
        bool firstTime = true;
        while (true)
        {
            for (int i = -10; i <= 10; i++)
            {
                doMaths(firstTime, i);
                doMaths(firstTime, i * 10000);
                doMaths(firstTime, i * 100000);
            }

            Waitframe();
            firstTime = false;
        }
    }

    void maybePrint(bool doPrint, char32[] str, int value)
    {
        if (doPrint)
        {
            printf(str, value);
        }
    }

    void doMaths(bool firstTime, int i)
    {
        maybePrint(firstTime, "maths for: %d\n", i);
        maybePrint(firstTime, "10 / i = %d", 10 / i);
        maybePrint(firstTime, "10 %% i = %d", 10 % i);
        maybePrint(firstTime, "i / 10 = %d", i / 10);
        maybePrint(firstTime, "i %% 10 = %d", i % 10);

		maybePrint(firstTime, "i + i = %d\n", i + i);
        maybePrint(firstTime, "i - i = %d\n", i - i);
        maybePrint(firstTime, "i * i = %d\n", i * i);
        maybePrint(firstTime, "i * -i = %d\n", i * -i);
        maybePrint(firstTime, "i / i = %d\n", i / i);
        maybePrint(firstTime, "i / -i = %d\n", i / -i);
        maybePrint(firstTime, "i %% (i - 50) = %d\n", i % (i - 50));

        maybePrint(firstTime, "i + 100000 = %d\n", i + 100000);
        maybePrint(firstTime, "i - 100000 = %d\n", i - 100000);
        maybePrint(firstTime, "i * 100000 = %d\n", i * 100000);
        maybePrint(firstTime, "i * -100000 = %d\n", i * -100000);
        maybePrint(firstTime, "i / 100000 = %d\n", i / 100000);
        maybePrint(firstTime, "i / -100000 = %d\n", i / -100000);
        maybePrint(firstTime, "i %% 50 = %d\n", i % 50);
        maybePrint(firstTime, "i %% -50 = %d\n", i % -50);
    }
}
