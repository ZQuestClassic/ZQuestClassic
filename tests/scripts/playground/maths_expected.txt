stderr:

Compiling 'maths.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

stdout:

{
  "success": true,
  "diagnostics": [],
  "metadata": {
    "currentFileSymbols": 6,
    "symbols": 14,
    "identifiers": 91,
    "elided": true
  }
}

zasm:

QUIT; void run() Body
PUSHR REFFFC; void run() Body Start
PUSHARGSV 0 0.0002
SETR D4 SP2
STOREV 1 0.0001
STOREV -10 0; for() #150 setup AND while(true) #2 Body [Opt:AlwaysOn] Start
LOAD D2 0; for() #150 LoopTest
COMPAREV D2 10
SETCMP D2 I<=
COMPAREV D2 0
GOTOCMP 37 ==; for() #150 TestFail?
PUSHR D4; for() #150 Body Start
LOAD D2 0.0001; UsrFunc[void doMaths(bool, int)] Params Start
PUSHR D2
LOAD D2 0
PUSHR D2; UsrFunc[void doMaths(bool, int)] Params End
CALLFUNC 0.0041; UsrFunc[void doMaths(bool, int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void doMaths(bool, int)] Params Start
PUSHR D2
LOAD D2 0
MULTV D2 10000
PUSHR D2; UsrFunc[void doMaths(bool, int)] Params End
CALLFUNC 0.0041; UsrFunc[void doMaths(bool, int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void doMaths(bool, int)] Params Start
PUSHR D2
LOAD D2 0
MULTV D2 100000
PUSHR D2; UsrFunc[void doMaths(bool, int)] Params End
CALLFUNC 0.0041; UsrFunc[void doMaths(bool, int)] Call
POP D4; for() #150 Body End
LOAD D2 0; for() #150 LoopIncrement
ADDV D2 1
STORE D2 0
GOTO 7; for() #150 End
WAITFRAME; InlineFunc[void Waitframe()] Body
SETV D2 0
STORE D2 0.0001; while(true) #2 Body [Opt:AlwaysOn] End
GOTO 6; while() #2 End AND [Opt:NeverRet] AND void run() Body End
SETR D4 SP2; Func[void doMaths(bool, int)] Body Start
PUSHR D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 15 0
WRITEPODSTRING D2 "maths for: %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "10 / i = %d"
PUSHR D2
LOAD D2 0
DIVV2 10 D2
SETV D2 10
PUSHR D2
LOAD D2 0
POP D3
DIVR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 13 0
WRITEPODSTRING D2 "10 %% i = %d"
PUSHR D2
LOAD D2 0
MODV2 10 D2
SETV D2 10
PUSHR D2
LOAD D2 0
POP D3
MODR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "i / 10 = %d"
PUSHR D2
LOAD D2 0
DIVV D2 10
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 13 0
WRITEPODSTRING D2 "i %% 10 = %d"
PUSHR D2
LOAD D2 0
MODV D2 10
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "i + i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
POP D3
ADDR D2 D3
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "i - i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
POP D3
SUBR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "i * i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
POP D3
MULTR D2 D3
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 13 0
WRITEPODSTRING D2 "i * -i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
SUBV2 0 D2
POP D3
MULTR D2 D3
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 12 0
WRITEPODSTRING D2 "i / i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
POP D3
DIVR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 13 0
WRITEPODSTRING D2 "i / -i = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
SUBV2 0 D2
POP D3
DIVR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 20 0
WRITEPODSTRING D2 "i %% (i - 50) = %d\n"
PUSHR D2
LOAD D2 0
PUSHR D2
LOAD D2 0
SUBV D2 50
POP D3
MODR D3 D2
PUSHR D3; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 17 0
WRITEPODSTRING D2 "i + 100000 = %d\n"
PUSHR D2
LOAD D2 0
ADDV D2 100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 17 0
WRITEPODSTRING D2 "i - 100000 = %d\n"
PUSHR D2
LOAD D2 0
SUBV D2 100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 17 0
WRITEPODSTRING D2 "i * 100000 = %d\n"
PUSHR D2
LOAD D2 0
MULTV D2 100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 18 0
WRITEPODSTRING D2 "i * -100000 = %d\n"
PUSHR D2
LOAD D2 0
MULTV D2 -100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 17 0
WRITEPODSTRING D2 "i / 100000 = %d\n"
PUSHR D2
LOAD D2 0
DIVV D2 100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 18 0
WRITEPODSTRING D2 "i / -100000 = %d\n"
PUSHR D2
LOAD D2 0
DIVV D2 -100000
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 14 0
WRITEPODSTRING D2 "i %% 50 = %d\n"
PUSHR D2
LOAD D2 0
MODV D2 50
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
PEEK D4
LOAD D2 0.0001; UsrFunc[void maybePrint(bool, char32[], int)] Params Start
PUSHR D2
ALLOCATEMEMV D2 15 0
WRITEPODSTRING D2 "i %% -50 = %d\n"
PUSHR D2
LOAD D2 0
MODV D2 -50
PUSHR D2; UsrFunc[void maybePrint(bool, char32[], int)] Params End
CALLFUNC 0.0278; UsrFunc[void maybePrint(bool, char32[], int)] Call
POP D4
POPARGS D5 0.0002
RETURNFUNC; Func[void doMaths(bool, int)] Body End
SETR D4 SP2; Func[void maybePrint(bool, char32[], int)] Body Start
MARK_TYPE_STACK 0.0001 0.0001
REF_INC 0.0001
LOAD D2 0.0002; if() #420 Test
COMPAREV D2 0
GOTOCMP 290 ==; Test 'if'
LOAD D2 0.0001; InlineFunc[void printf(char32[])] Params Start AND if() #420 Body Start
PUSHR D2
LOAD D2 0
PUSHVARGR D2; InlineFunc[void printf(char32[])] Params End
PRINTFVARG; InlineFunc[void printf(char32[])] Body Start
POP D5; InlineFunc[void printf(char32[])] Body End AND if() #420 Body End
REF_REMOVE 0.0001
POPARGS D5 0.0003
RETURNFUNC; Func[void maybePrint(bool, char32[], int)] Body End
